python/ray/dag/compiled_dag_node.py

```diff
@@ -843,32 +896,52 @@ def _get_or_compile(
                     #     print(result)
 
                     #     compiled_dag.teardown()
-                    reader_handles = [self._driver_actor]
+                    assert (self._driver_actor is not None) ^ (
+                        self._actor_handle is not None
+                    )
+                    if self._driver_actor is not None:
+                        reader_to_node.append(
+                            (self._driver_actor, self._get_node_id(self._driver_actor))
+                        )
+                    else:
+                        reader_to_node.append(
+                            (self._actor_handle, self._get_node_id(self._actor_handle))
+                        )
                 else:
-                    reader_handles = [
-                        reader.dag_node._get_actor_handle() for reader in readers
-                    ]
+                    for reader in readers:
+                        reader_handle = reader.dag_node._get_actor_handle()
+                        reader_to_node.append(
+                            (reader_handle, self._get_node_id(reader_handle))
+                        )
+
                 fn = task.dag_node._get_remote_method("__ray_call__")
                 task.output_channel = ray.get(
                     fn.remote(
                         do_allocate_channel,
-                        reader_handles,
+                        reader_to_node,
                         typ=type_hint,
                     )
                 )
                 actor_handle = task.dag_node._get_actor_handle()
                 self.actor_refs.add(actor_handle)
                 self.actor_to_tasks[actor_handle].append(task)
             elif isinstance(task.dag_node, InputNode):
+                reader_to_node: List[Tuple["ray.actor.ActorHandle", str]] = []
+                # TODO (kevin85421): We need to remove reader_handles_set because
+                # readers can be on the same actor.
                 reader_handles_set = set()
                 for idx in task.downstream_node_idxs:
                     reader_task = self.idx_to_task[idx]
                     assert isinstance(reader_task.dag_node, ClassMethodNode)
                     reader_handle = reader_task.dag_node._get_actor_handle()
+                    if reader_handle not in reader_to_node:
```

Comment:
Looks like the type does not match? reader_handle is an actor handle, reader_to_node is a list of tuples

Reply:
Good catch. I initially used a dictionary, but then changed to a list and forgot to update it.